File: src/native/network.ts
import ffi, { cstr } from "./ffi.ts";
import { Layer } from "./layer.ts";
import { Matrix } from "./matrix.ts";

const {
  network_free,
  network_create,
  network_feed_forward,
  network_train,
  network_load,
  network_save,
} = ffi;

const NetworkFinalizer = new FinalizationRegistry(
  (network: Deno.PointerValue) => {
    network_free(network);
  },
);

enum C_COST {
  crossentropy = 0,
  mse = 1,
}

export type Cost = keyof typeof C_COST;

export interface NetworkConfig {
  inputSize: number;
  layers: Layer[];
  cost: Cost;
}

export interface Dataset {
  inputs: Matrix<"f32">;
  outputs: Matrix<"f32">;
}

export class Network {
  #ptr: Deno.PointerValue;
  #token: { ptr: Deno.PointerValue } = { ptr: 0 };

  get unsafePointer() {
    return this.#ptr;
  }

  constructor(config: NetworkConfig | Deno.PointerValue) {
    this.#ptr = typeof config === "object"
      ? network_create(
        config.inputSize,
        C_COST[config.cost],
        config.layers.length,
        new BigUint64Array(config.layers.map((e) => BigInt(e.unsafePointer))),
      )
      : config;
    this.#token.ptr = this.#ptr;
    NetworkFinalizer.register(this, this.#ptr, this.#token);
  }

  predict(input: Matrix<"f32">): Matrix<"f32"> {
    return new Matrix(network_feed_forward(this.#ptr, input.unsafePointer));
  }

  train(datasets: Dataset[], epochs: number, learningRate: number) {
    const datasetBuffers = datasets.map((e) =>
      new BigUint64Array(
        [e.inputs.unsafePointer, e.outputs.unsafePointer].map(BigInt),
      )
    );
    const datasetBufferPointers = new BigUint64Array(
      datasetBuffers.map((e) => BigInt(Deno.UnsafePointer.of(e))),
    );
    network_train(
      this.unsafePointer,
      datasets.length,
      datasetBufferPointers,
      epochs,
      learningRate,
    );
  }

  static load(path: string): Network {
    return new Network(network_load(cstr(path)));
  }

  save(path: string) {
    network_save(this.#ptr, cstr(path));
  }

  free(): void {
    if (this.#ptr) {
      network_free(this.#ptr);
      this.#ptr = 0;
      NetworkFinalizer.unregister(this.#token);
    }
  }
}

File: src/network.rs
use ndarray::ArrayD;

use crate::{Backend, BackendType, CPUBackend, Layer, Cost, Dataset};

pub struct NetworkConfig {
    pub size: &'static [usize],
    pub layers: Vec<Layer>,
    pub backend: BackendType,
    pub cost: Cost
}

pub struct Network {
    pub backend: Backend,
}

impl Network {
    pub fn new(config: NetworkConfig) -> Self {
        let backend = match config.backend {
            BackendType::CPU => Backend::CPU(CPUBackend::new(config)),
        };
        Self { backend }
    }

    pub fn train(&mut self, datasets: Vec<Dataset>, epochs: usize, rate: f32) {
        match &mut self.backend {
            Backend::CPU(backend) => backend.train(datasets, epochs, rate)
        }
    }

    pub fn predict(&mut self, data: ArrayD<f32>) -> ArrayD<f32> {
        match &mut self.backend {
            Backend::CPU(backend) => backend.predict(data)
        }
    }
}
